lib/src/png/png_metadata.dart

```diff
@@ -19,53 +19,95 @@ extension PngMetadata on PngData {
       final dataEnd = dataStart + dataLength;
 
       // Move data offset forward.
-      offset = dataEnd + 4; // +4 to skip CRC
+      offset += 12 + dataLength; // length + type + data + CRC
 
       if (chunkType != 'iTXt') {
         continue;
       }
 
-      final data = sublist(dataStart, dataEnd);
-      final separator = data.indexOf(0);
-      if (separator <= 0) {
-        continue;
+      final chunkData = sublist(dataStart, dataEnd);
+
+      // Parse iTXt chunk format
+      final reader = _ChunkReader(chunkData);
+      final keyword = reader.readNullTerminatedString();
+      final compressionFlag = reader.readByte();
+      // ignore: unused_local_variable
+      final compressionMethod = reader.readByte();
+      // ignore: unused_local_variable
+      final languageTag = reader.readNullTerminatedString();
+      // ignore: unused_local_variable
+      final translatedKeyword = reader.readNullTerminatedString();
+      final textBytes = reader.readRemaining();
+
+      if (compressionFlag != 0) {
+        throw Exception(
+          "Tried to read PNG but one if its iTXt fields is compressed. flutter_test_goldens does not currently support reading/writing compressed text.",
+        );
       }
+      final text = utf8.decode(textBytes);
 
-      result[utf8.decode(data.sublist(0, separator))] = utf8.decode(data.sublist(separator + 1));
+      result[keyword] = text;
     }
 
     return result;
   }
 
   Uint8List copyWithTextMetadata(String key, String value) {
-    final keyUtf8 = utf8.encode(key);
+    final png = Uint8List.fromList(this);
+    final output = BytesBuilder();
 
-    final valueUtf8 = utf8.encode(value);
-    final valueOffset = keyUtf8.length + 1;
+    // Write PNG header (8 bytes)
+    output.add(png.sublist(0, 8));
 
-    final totalLength = valueOffset + valueUtf8.length;
+    // Iterate through chunks
+    int offset = 8;
+    while (offset < png.length) {
+      final length = _readUint32(offset);
+      final type = utf8.decode(png.sublist(offset + 4, offset + 8));
```

Comment:
Just a note. The tags are encoded in ASCII, but utf8 decode is compatible with that.